californium-core/src/main/java/org/eclipse/californium/core/network/Exchange.java

```diff
@@ -583,10 +586,13 @@ public void setCustomExecutor() {
 	 * @param ctx the correlation information
 	 */
 	public void setCorrelationContext(final CorrelationContext ctx) {
-		correlationContext = ctx;
-		ExchangeObserver obs = this.observer;
-		if (obs != null) {
-			obs.contextEstablished(this);
+		if (correlationContext.compareAndSet(null, ctx)) {
+			ExchangeObserver obs = this.observer;
+			if (obs != null) {
+				obs.contextEstablished(this);
+			}
+		} else {
+			correlationContext.set(ctx);
```

Comment:
https://github.com/eclipse/californium/issues/311#issuecomment-306852749

> My response is "no, it must not change".

https://github.com/eclipse/californium/issues/311#issuecomment-306860252

> I do not see how it could, actually. Am I missing something?


https://github.com/eclipse/californium/issues/311#issuecomment-306860933

> It could be changed each time we send a blockwise request. (If I read the code correctly)

So I don't see, that this discussion is finished.

https://github.com/eclipse/californium/issues/311#issuecomment-306839999

> But I would feel more comfortable, if the discussion in the mailing list came to a conclusion.

That's still my position. Change enough to optimize the "critical callback", but less as possible while it is discussed.